services/elasticache/pkg/resource/replication_group/custom_update_api.go

```diff
@@ -30,76 +30,146 @@ func (rm *resourceManager) CustomModifyReplicationGroup(
 	latest *resource,
 	diffReporter *ackcompare.Reporter,
 ) (*resource, error) {
-	for _, diff := range diffReporter.Differences {
-		switch diff.Path {
-		case "Spec.NumNodeGroups":
-			return rm.updateShardConfiguration(ctx, desired, latest, diffReporter)
-		case "Spec.ReplicasPerNodeGroup":
-			return rm.updateReplicaCount(ctx, desired, latest, diffReporter)
+	// Order of operations when diffs map to multiple updates APIs:
+	// 1. updateReplicaCount() is invoked Before updateShardConfiguration()
+	//	  because both accept availability zones, however the number of
+	//	  values depend on replica count.
+
+	// increase/decrease replica count
+	if diff, message := rm.replicaCountDifference(desired, latest); diff != 0 {
+		fmt.Print(message)
+		if diff > 0 {
+			return rm.increaseReplicaCount(ctx, desired, latest)
 		}
+		return rm.decreaseReplicaCount(ctx, desired, latest)
+	}
+
+	// increase/decrease shards
+	if rm.shardConfigurationsDiffer(desired, latest) {
+		return rm.updateShardConfiguration(ctx, desired, latest)
 	}
+
+	// no updates
 	return nil, nil
 }
 
-func (rm *resourceManager) updateShardConfiguration(
+// replicaConfigurationsDifference returns
+// +ve number if desired replica count is greater than latest replica count
+// -ve number if desired replica count is less than latest replica count
+// 0 otherwise
+func (rm *resourceManager) replicaCountDifference(
+	desired *resource,
+	latest *resource,
+) (int, string) {
+	desiredSpec := desired.ko.Spec
+	latestStatus := latest.ko.Status
+
+	if desiredSpec.ReplicasPerNodeGroup != nil {
+		for _, latestShard := range latestStatus.NodeGroups {
+			latestReplicaCount := 0
+			for _, latestShardMember := range latestShard.NodeGroupMembers {
+				if latestShardMember.CurrentRole != nil && *latestShardMember.CurrentRole == "replica" {
+					latestReplicaCount++
+				}
+			}
+			if desiredReplicaCount := int(*desiredSpec.ReplicasPerNodeGroup); desiredReplicaCount != latestReplicaCount {
+				nodeGroupID := ""
+				if latestShard.NodeGroupID != nil {
+					nodeGroupID = *latestShard.NodeGroupID
+				}
+				return desiredReplicaCount - latestReplicaCount, fmt.Sprintf(
+					"ReplicasPerNodeGroup differs for NodeGroup: %s, desired = %d, latest: %d",
+					nodeGroupID, int(*desiredSpec.ReplicasPerNodeGroup), latestReplicaCount)
+			}
+		}
```

Comment:
might be nice to break the above lines 68-84 into its own `diffReplicasPerNodeGroup` function and lines 86-115 into its own `diffReplicasNodeGroupConfiguration` function that you can then call:

```go
if desiredSpec.ReplicasPerNodeGroup != nil {
    return rm.diffReplicasPerNodeGroup(desired, latest)
} else if desiredSpec.NodeGroupConfiguration != nil {
    return rm.diffReplicasNodeGroupConfiguration(desired, latest)
}
return 0
```

Reply:
Thanks for the suggestion. Updated source code.